### PR TITLE
Implement allPairs on List, Seq and Array modules

### PR DESCRIPTION
**Pull request by PatrickMcDonald**
This is an initial pull request that implements allPairs on the collections.

See fsharp/FSharpLangDesign#47 and https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1002-cartesian-product-for-collections.md

### CODE REVIEW

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/ListProperties.fs [hunk 1]
```diff
@@ -823,3 +823,35 @@ let ``List.sumBy calculates the sum of the mapped list`` () =
...
+    pairsFst = check
+
+[<Test>]
+let ``List.allPairs first elements are correct`` () =
+    Check.QuickThrowOnFailure allPairsFst<int, int>
+    Check.QuickThrowOnFailure allPairsFst<string, string>
+    Check.QuickThrowOnFailure allPairsFst<NormalFloat, NormalFloat>
+
+let allPairsSnd<'a, 'b when 'b : equality> (xs : 'a list) (ys : 'b list) =
+    let pairsFst = List.allPairs xs ys |> List.map snd
```
Comment (mexx) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/ListProperties.fs]: can you please rename it to `pairsSnd`

Comment (PatrickMcDonald) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/ListProperties.fs]: Thanks

File: src/fsharp/FSharp.Core/array.fsi [hunk 2]
```diff
@@ -12,6 +12,14 @@ namespace Microsoft.FSharp.Collections
     [<RequireQualifiedAccess>]
     module Array = 
 
+        /// <summary>Builds a new array that contains the cartesian product of the two input arrays.</summary>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fsi]: Please use a less technical term than "Cartesian product" in the comment, such as "contains all pairings of elements from the first and second arrays", likewise the other three signature file comments

**Comment by dsyme** [comment]
This looks good to me and ready to go apart from the one documentation comment above (which could perhaps be applied later)

**Comment by KevinRansom** [comment]
I will take this PR as is.  @PatrickMcDonald.

**Comment by hickford** [comment]
Both https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/lists and https://msdn.microsoft.com/visualfsharpdocs/conceptual/collections.list-module-%5bfsharp%5d are missing documentation for List.allPairs

**Pull request merged**